doc/arch/adr-010-learn-to-rank.md

```diff
@@ -0,0 +1,286 @@
+
+This validated for us that Learning To Rank is feasible and could
+lead to more relevant results for users.
+
+Though Learning To Rank won't make relevance tuning easier, it could
+make search results more relevant.
+
+### Final decision
+
+We feel confident about implementing Learning To Rank (LTR) for the
+following reasons:
+
+- We think that it will improve search relevancy for citizens, enabling
+  them to find what they need on GOV.UK
+- We have validated our assumptions with a spike
+- LTR is a well trodden path, and other organisations have written about
+  their implementations with varying levels of openness (e.g. Wikimedia)
+- GOV.UK has implemented other machine learning projects with success
+  (related links, topic taxonomy), and we have a data science team which
+  the search team has already worked with
+- TensorFlow provides an off-the-shelf ranking module, which solves
+  some of the machine learning problems of LTR for us
+- Manual relevancy tuning is not scalable, and won't get us to the same
+  level of relevancy as LTR
+
+## Implementation plan
+
+We have decided to implement Learning To Rank by following on from
+the work started in the [validation spike][].
+
+As we are at the beginning of the project, this section may not
+accurately reflect the final implementation.
+
+[validation spike]: https://github.com/alphagov/search-api/pull/1768
+
+### Language
+
+We would like to keep the code for this project in Ruby as much as
+possible. Search API is written in Ruby and it is the primary language
+used at GOV.UK. Necessarily some machine learning code will need to be
+written in Python.
+
+### Machine learning library
+
+We have decided to [TensorFlow's ranking library][] for our
+implementation of Learning To Rank.
+
+TensorFlow is well supported, and the ranking library is under active development.
+
+We considered using the [elasticsearch-learning-to-rank][] plugin.
+However, AWS Elasticsearch does not permit us to install custom plugins.
+
+We also considered PyTorch. However, this would require writing much more
+custom machine learning code, as PyTorch don't currently have a ranking library.
+
+[TensorFlow's ranking library]: https://github.com/tensorflow/ranking
+[elasticsearch-learning-to-rank]: https://github.com/o19s/elasticsearch-learning-to-rank
+
+### Dockerised model
+
+We will use TensorFlow Serving to make a re-ranker model available
+to Search API on search machines.
+
+We will run a docker container (containing our pre-trained model) on
+search machines to which Search API can send requests.
+
+Should the docker container fail, Search API should be resilient to this,
+and fall back to the default ranking algorithm, with results not
+re-ranked by the model.
+
+### Training
+
+**Training the model** should happen on search machines on a schedule.
+An instance of Search API will train a model using the latest relevancy
+judgements stored in S3, and will then upload a new version of the model to S3.
```

Comment:
@karlbaker02 's approach was I think to spin up a temporary EC2 instance for generating models using a concourse pipeline.  Is this solving a different problem?

Reply:
Very good point. It might be nice to do have a separate instance, not responsible for responding to user requests, do the training.

It might be that the model doesn't need a long training time (e.g. 10 minutes) so doing that out-of-hours once a day won't require a specialised machine. On the other hand if it'll be a multi-hour highly compute intensive job I can see the value of using specialised instance.

At the moment I am not sure about the compute / time requirements of training the model, so we could delay this decision. What do you think?

Reply:
Our model for related links originally took around 9hrs to complete, which has been reduced in half since through various improvements; this, combined with the need for a different instance type was the main reason we opted for a temporary EC2 instance managed via Concourse. Of course, this was a new project so there wasn't an existing box which might have made sense to put this on, something you have in this case as it's directly related to search.

If the model doesn't take long to generate then it could be run on an existing machine (maybe out-of-hours to make sure we don't compromise the machine's resources when search might be being used more)? 

Another option you could consider is AWS Lambda, so long as your process completes within 15 minutes (the max running time allowed).

Reply:
I don't know what the infrastructure is like now, so take this with a grain of salt, but this part stood out to me as well. I think that GOV.UK's use of ML will only grow over time, and there's a risk of building up a lot of infrastructure that does similar things in different ways. The more data engineering there is on GOV.UK the more likely it is you will need to add new processing steps, consume data from different places, or reuse code for other purposes. 

In this case I don't think there's a need for model building to happen on the same machines that serve requests, and if you can isolate those two things (serving requests vs training the model) the processes will be more resiliant and you can scale them independently.

I haven't used concourse before, but if you have it set up already, and it works, that sounds much better than building the data engineering from scratch. I'm imagining this pipeline will have multiple steps, to clean data, train the model, test it, and deploy it. There are lots of tools for this kind of thing (I know verify have built a bunch of data pipelines in luigi) but I think the most important thing is that it's easier to maintain than the current system of tweaking magic numbers and things!

I also strongly recommend python for this side of things because you can count on data scientists and data engineers knowing it and it comes with a massive community and ecosystem.

Reply:
Generating a new model is definitely becoming slower since we started the project. Previously we could generate a new model in under an hour, because we were dealing with small datasets.

Now we spend a few minutes fetching data from big query, followed by 2-3 hours of generating relevancy judgements from that data, and then a further 1-2 hours training the model. The time it takes to generate training sets and train the model grows with the number of days of user data that we use. There are likely a lot of performance improvements that could be made in this code, but it's likely going to be fairly slow!

It might be that we could run this process once a week out of hours. But if we wanted to try out new models frequently in-hours it might be nice to have a dedicated machine for this.

Running it on concourse sounds good. @karlbaker02 let's chat about this! @MatMoore is probably right about us duplicating data pipeline stuff for this work, I'm interested in what the data labs team has already developed that we could re-use.

Reply:
I agree with everything @MatMoore said above

A few points from me:

- Choosing Concourse for Related Links was mostly because it was there and was supported by RE (and still is supported). Now that we have more of this type of work we have a chance to reconsider this particular choice for data pipelines. We are currently looking at various options including AWS Glue for ETL type of jobs
- It doesn’t feel right hosting a docker image with trained model on search boxes. And if we end up using docker containers  we need to get RE’s view on hosting them (Fargate?). We might even need to consider PaaS for this. Another option I am researching at the moment is AWS EMR, which is a serverless cluster that can run Tensorflow tasks among other things
- I am very sceptical about AWS Lambda for this as it also have RAM, storage space and other limitations